[RPC] require valid URL scheme on budget commands

CaveSpectre11 opened the pull request:
### **Release Notes** 
[RPC] Require valid URL schemes (URLs starting with "http" or "https") on preparebudget and submitbudget.

### **Details** 
The budget proposal system did not have any checks on the URL aside of string length.  This allowed for the scheme to be left off the URL; which on some browsers, resulted in the link to fail when clicked:

![image](https://user-images.githubusercontent.com/36988814/61176879-69c97a80-a596-11e9-9307-cbeeb210e133.png)
```
gio: file:///home/CaveSpectre/forum.pivx.org/index.php%3Fthreads/bizdev-proposal-john-m.285: 
Error when getting information for file “/home/CaveSpectre/forum.pivx.org/index.php?
threads/bizdev-proposal-john-m.285”: No such file or directory
```

The included PR changes the `submitbudget` and `preparebudget` rpc commands to prevent inadvertent omissions of the URL scheme (http:// or https://).  

Qt changes to fixup the existing & historical budget URLs will be submitted as a separate PR after the merging of #954 .

![image](https://user-images.githubusercontent.com/36988814/61176955-017b9880-a598-11e9-8df0-f4afe1ccdf7d.png)

_(This PR originally included the Qt changes, and as a result some of the below discussion relates to that, before the decision was made to back out the Qt changes until a later date)_

Warrows commented on src/masternode-budget.h:
```diff
@@ -515,6 +520,17 @@ class CBudgetProposal
+        std::string ret=strURL;
+        if (!regex_match(strURL, URLRegEx))
+        {
+           ret = scheme+"://"+strURL;
```
nit: missing 1 space at the beginning of the line.

CaveSpectre11 replied:
corrected; rebased and squashed (I think :)

CaveSpectre11 commented:
I originally thought that regex failure in the zerocoin test was something unrelated; or related to the zerocoin changes.  But now that I look at it.... 

@Warrows , do you happen to know why that test uses gcc 4.8?  regex should be stable in 4.9, and most of the builds appear to be using 7.4.  Are we restricted to supporting 4.8?

Warrows commented:
That's a question better suited for @fuzzbawls

CaveSpectre11 commented:
I should be able to conditionalize it so that `regex_match()` just returns true on builds prior to gcc 4.9, which would effectively fix the travis test, but have no impact on the releases.   I will wait for @Fuzzbawls to suggest direction.

```
#include <regex>

#if __cplusplus >= 201103L &&                             \
    (!defined(__GLIBCXX__) || (__cplusplus >= 201402L) || \
        (defined(_GLIBCXX_REGEX_DFS_QUANTIFIERS_LIMIT) || \
         defined(_GLIBCXX_REGEX_STATE_LIMIT)           || \
             (defined(_GLIBCXX_RELEASE)                && \
             _GLIBCXX_RELEASE > 4)))
#  define HAVE_WORKING_REGEX 1
#else
#  define HAVE_WORKING_REGEX 0
#endif
```

It'll take a little toying to get it right

Fuzzbawls commented:
Concept ACK, but would need to do this without relying on `std::regex` so as to avoid bumping the minimum supported ABI.

maybe have a look at using `std::string::find` instead.

CaveSpectre11 commented:
> std::string::find

It wouldn't be as robust; but yes, searching for "://" would at least solve the inadvertent and common omission.

CaveSpectre11 commented:
Quick test log of the rpc restrictions.  logically, "h://" and "H://" pass, but we at least force [hH]+:// which should make it better.

akshaynexus commented on src/masternode-budget.h:
```diff
@@ -515,6 +515,19 @@ class CBudgetProposal
     void CleanAndRemove(bool fSignatureCheck);
 
+    // If URL doesn't have a scheme, add one
+    std::string GetFixedupURL(std::string scheme="http")
```
nit,might be better having scheme as https instead of http

CaveSpectre11 replied:
So here's my logic for choosing http rather than https:

1. Not all servers have appropriate https (although it's may kick you to http; I didn't test).
2. If they wanted to be specific, they could have specified it with the proposal.
3. It does pop up a window showing them the URL and asking them to confirm it before continuing; so they can copy it and replace it with https.

But this portion (related to Qt) may be changed with the new wallet anyway.

So: noted; and I'll see what comes with the new wallet and fuzzballs direction (as we are in a holding pattern with this for now)

random-zebra commented:
I don't particularly like this solution.
Think it would be better to use `tolower()` (even if locale dependent and requiring some additional changes in the lint script) on everything before `://` and compare with the allowed prefixes.

Anyway, would code the logic in a separate utility function and call that from `submitbudget` and `preparebudget` to avoid more code repetition (those two functions already share way too much code in my opinion).

CaveSpectre11 commented:
> Anyway, would code the logic in a separate utility function and call that from `submitbudget` and `preparebudget` to avoid more code duplication (those two functions already share way too much code in my opinion).

The original design had a utility function for adding the scheme; but the code check in the rpc code was less complicated.  I think I agree that a utility function is the way to go; especially given the premise of the last round of changes (adding more URL scheme) validation.   Pulling it together into a utility function centralizes where future code can go, if the need to have further validation of the URL is desired later.

> Think it would be better to use `tolower()` (even if locale dependent and requiring some additional changes in the lint script) on everything before `://` and compare with the allowed prefixes.

I will go ahead and pull the logic into a function; and you and @Fuzzbawls can hash out the pros and cons of `tolower()` and finish it out when that direction is determined.

CaveSpectre11 commented:
@Fuzzbawls @random-zebra; I have submitted a much broader PR (#964 ) that encompasses this PR, with modifications to the logic to hopefully account for both your concerns.

That PR does supersede this one, and if that solution is preferred, this one should be closed.

CaveSpectre11 commented:
Changes in the last push:
- rebased on top of #964 (refactoring) for ease of rebasing later
- moved the routine to utilstrencoding.h
- added the URL length as a parameter, defaulting to 64  (this can be expanded on in the future to allow for the search vectors to be tweaked by the caller)
- implemented the last discussed rules set; requiring for lowercase "http://" or "https://"

CaveSpectre11 commented:
> ```c++
> if(!validateURL(strURL, strErr))
> 	throw JSONRPCError(RPC_INVALID_PARAMETER, strErr);
> ```

Done

random-zebra commented on src/rpc/budget.cpp:
```diff
@@ -48,14 +48,53 @@ void budgetToJSON(CBudgetProposal* pbudgetProposal, UniValue& bObj)
+    strURL = SanitizeString(params[1].get_str());
+    std::string strErr;
+    if (!validateURL(strURL, strErr))
+        throw JSONRPCError(RPC_INVALID_PARAMETER, result);
```
`result` is undefined

CaveSpectre11 replied:
oh crap; that's what happens when I make "one last change!" while my kids are trying to get my attention :)

random-zebra replied:
I can totally relate :stuck_out_tongue_closed_eyes:

CaveSpectre11 replied:
cleaned up and build tested

random-zebra commented on src/rpc/budget.cpp:
```diff
@@ -48,14 +48,53 @@ void budgetToJSON(CBudgetProposal* pbudgetProposal, UniValue& bObj)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid block start, must be more than current height.");
+
+    if (nBlockEnd < pindexPrev->nHeight)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid ending block, starting block + (payment_cycle*payments) must be more than current height.");
```
Isn't this unreachable code?
If we are here it means 
```
nBlockEnd = nBlockStart + (...) > nBlockStart  >= nBlockMin >= pindexPrev->nHeight
```
thus `nBlockEnd > pindexPrev->nHeight`

CaveSpectre11 replied:
Correct; these are the things I saw when I got deeper into this and thus #965 turned up.

random-zebra replied:
Oh, right. Guess I'm getting confused with the order of these PRs.
Shouldn't 965 be rebased on top of this one then?

CaveSpectre11 replied:
Trust me i keep getting confused with them all too.   I can merge that one in here so this can be the final picture, but have the 3 separate commits to reflect the distinctions until they start going into mainline

random-zebra replied:
That would be fine by me.

Other solutions could be
1) rebase 965 on top of this one (and have merge order 964 --> 950 --> 965)
2) rebase this one on top of 965 (and merge like 964 --> 965 --> 950)

CaveSpectre11 replied:
#2 is pretty much the  likely scenario; since that -should- be the merge order.

random-zebra commented on src/rpc/budget.cpp:
```diff
@@ -48,14 +48,53 @@ void budgetToJSON(CBudgetProposal* pbudgetProposal, UniValue& bObj)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid payment count, must be more than zero.");
+
+    // Start must be in the next budget cycle
+    if (pindexPrev != NULL) nBlockMin = pindexPrev->nHeight - pindexPrev->nHeight % Params().GetBudgetCycleBlocks() + Params().GetBudgetCycleBlocks();
```
nit: use `nullptr`

CaveSpectre11 replied:
Also already addressed in #965

random-zebra commented on src/rpc/budget.cpp:
```diff
@@ -48,14 +48,53 @@ void budgetToJSON(CBudgetProposal* pbudgetProposal, UniValue& bObj)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid block start - must be a budget cycle block. Next valid block: %d", nNext));
+    }
+
+    int nBlockEnd = nBlockStart + (Params().GetBudgetCycleBlocks() * nPaymentCount); // End must be AFTER current cycle
```
nit: this can be set after the `if (nBlockStart < nBlockMin)` check.

CaveSpectre11 replied:
actually; nBlockEnd is useless and was removed in 965; classic high school transitive property problem; if a < b and b < c then a < c and don't need to waste the cycles to prove the property.

CaveSpectre11 commented:
Ok; I just totally jacked this one up trying to merge in #965.  I'll try to figure it out tomorrow.

Mrs-X commented:
Just some random thoughts:
- `checkBudgetInputs()` was already there, so why did you copy that code into the methods?
- the canonical form of a scheme is lowercase, so I second forcing `http://` or `https://`

CaveSpectre11 commented:
> * `checkBudgetInputs()` was already there, so why did you copy that code into the methods?

#964 was meant as a replacement to this one originally when I submitted it.  But it was decided they should be separate.  This one to deal with the scheme check, 964 for the refactor only; and 965 for the fixes on the existing code.  The refactoring was what combined the common code in the prepare and submit into `checkBudgetInputs()`.

Then to make it more confusing; 964's change was cherry picked into 965; and both were picked into this one... presuming the merge order to mainline would be 964 first; then 965, then this one... and thus the rebasing legwork was already done.

CaveSpectre11 commented:
rebased due to conflict from on budget.cpp

random-zebra commented:
Needs rebase after #965 merge.

CaveSpectre11 commented:
> Needs rebase after #965 merge.

Rebased through github; will squash that commit when I get home.

Fuzzbawls commented on src/utilstrencodings.cpp:
```diff
@@ -39,6 +39,46 @@ std::string SanitizeString(const std::string& str, int rule)
+** Return:
+**    boolean true|false result if the validation passes.
+**    strRrror: Filled with any error messages.
+*/
```
Move this description comment block to the header file and format it according to the doxygen-compatible comment guidelines.

CaveSpectre11 replied:
complete.

random-zebra commented on src/utilstrencodings.cpp:
```diff
@@ -39,6 +39,33 @@ std::string SanitizeString(const std::string& str, int rule)
+    // check fronts
+    bool found = false;
+    for (int i=0; i < reqPre.size() && !found; i++) {
+        if (strURL.find(reqPre[i]) == 0) found = true;
```
minor nit: it would be nice to add a `break` after `found = true`.
Not really something that needs to be addressed now, since we have to check just for two cases (but something to keep in mind if later the prefixes list gets expanded).

CaveSpectre11 replied:
Ahh yes; good catch.

random-zebra replied:
whops nevermind... I see there's `i < reqPre.size() && !found;` in the cycle condition. :)

CaveSpectre11 replied:
lol!  I did find it surprising that I would have let something like that slip by.  I guess I didn't :)
